vm/interpreter/src/lib.rs

```diff
@@ -1,130 +1,319 @@
+                    "Sender state invalid".to_owned(),
+                )),
+            ));
+        };
 
-        // TODO: verify nonce of the from actor matches nonce of the message
+        let gas_cost = msg.gas_price() * msg.gas_limit();
+        // TODO requires network_tx_fee to be added as per the spec
+        let total_cost = &gas_cost + msg.value();
+        if from_act.balance < total_cost {
+            return Ok(ApplyRet::new(
+                MessageReceipt {
+                    return_data: Serialized::default(),
+                    exit_code: ExitCode::SysErrSenderStateInvalid,
+                    gas_used: BigUint::zero(),
+                },
+                miner_penalty_amount,
+                Some(ActorError::new(
+                    ExitCode::SysErrSenderStateInvalid,
+                    "Sender state invalid".to_owned(),
+                )),
+            ));
+        };
 
-        // TODO: check that the from actor has enough gas for the total gas cost
+        self.state.mutate_actor(msg.from(), |act| {
+            from_act.deduct_funds(&gas_cost)?;
+            act.sequence += 1;
+            Ok(())
+        })?;
 
-        // TODO: transfer gas
+        let snapshot = self.state.snapshot()?;
 
-        // TODO: increase from actor nonce
+        // scoped to deal with mutable reference borrowing
+        let (ret_data, gas_used, act_err) = {
+            let (ret_data, mut rt, act_err) = self.send(msg, msg_gas_cost as i64);
+            rt.charge_gas(rt.price_list().on_chain_return_value(ret_data.len()))
+                .map_err(|e| e.to_string())?;
+            (ret_data, rt.gas_used(), act_err)
+        };
 
-        let return_data = self.send(msg, gas_cost.to_i64().unwrap());
-        return_data
-            .map(|r| {
-                Ok(MessageReceipt {
-                    return_data: r, // TODO: what about Send?,
-                    exit_code: ExitCode::Ok,
-                    gas_used: BigUint::from(0u64), // TODO: get from runtime, runtime.gas_used()
-                })
-            })
-            .map_err(|e| {
+        if let Some(err) = act_err {
+            if err.is_fatal() {
+                return Err(format!("Fatal send actor error occurred, err: {:?}", err));
+            };
+            if err.exit_code() != ExitCode::Ok {
+                // revert all state changes since snapshot
                 if let Err(state_err) = self.state.revert_to_snapshot(&snapshot) {
-                    return state_err;
-                }
-                e.to_string()
-            })?
+                    return Err(format!("Revert state failed: {}", state_err));
+                };
+            }
+            warn!("Send actor error: from:{}, to:{}", msg.from(), msg.to());
+        }
+        let gas_used = if gas_used < 0 { 0 } else { gas_used as u64 };
+        // refund unused gas
+        let refund = (msg.gas_limit().checked_sub(gas_used).unwrap_or(1u64)) * msg.gas_price();
```

Comment:
```suggestion
        let refund = (msg.gas_limit().checked_sub(gas_used).unwrap_or(0u64)) * msg.gas_price();
```
Is checked sub actually necessary here? Don't we have a guarantee that gas used cannot be greater than the gas limit?

Reply:
I cannot find this guarantee anywhere. Do you have a reference?